### PR TITLE
(TWILL-188) Added methods to Location for manipulation of permission

### PR DESCRIPTION
**Pull request by chtyim**
- Three methods are added
  void createNew(String permission)
  String getPermissions()
  void setPermissions(String permission)

### CODE REVIEW

File: twill-common/src/main/java/org/apache/twill/filesystem/LocalLocation.java [hunk 1]
```diff
@@ -255,4 +288,53 @@ public int hashCode() {
...
+  /**
+   * Parses a three digits UNIX numeric permission representation to a set of {@link PosixFilePermission}.
+   */
+  private Set<PosixFilePermission> parseNumericPermissions(String permission) {
+    String posixPermission = "";
+    for (int i = 0; i < 3; i++) {
+      int digit = permission.charAt(i) - '0';
+      if (digit < 0 || digit > 7) {
+        throw new IllegalArgumentException("Invalid permission " + permission +
+                                             ". Only digits between 0-7 is allowed.");
```
Comment (albertshau) [inline twill-common/src/main/java/org/apache/twill/filesystem/LocalLocation.java]: is -> are

File: twill-common/src/main/java/org/apache/twill/filesystem/LocalLocation.java [hunk 2]
```diff
@@ -255,4 +288,53 @@ public int hashCode() {
...
+   */
+  private Set<PosixFilePermission> parsePermissions(String permission) {
+    Set<PosixFilePermission> permissions;
+    if (permission.length() == 3) {
+      permissions = parseNumericPermissions(permission);
+    } else if (permission.length() == 9) {
+      permissions = PosixFilePermissions.fromString(permission);
+    } else {
+      throw new IllegalArgumentException("Invalid permission " + permission +
+                                           ". Permission should either be a three digits or nine characters string.");
```
Comment (albertshau) [inline twill-common/src/main/java/org/apache/twill/filesystem/LocalLocation.java]: digits -> digit

File: twill-common/src/main/java/org/apache/twill/filesystem/Location.java [hunk 3]
```diff
@@ -53,11 +54,42 @@
    * Atomically creates a new, empty file named by this abstract pathname if and only if a file with this name
    * does not yet exist.
    * @return {@code true} if the file is successfully create, {@code false} otherwise.
-   * @throws IOException
+   * @throws IOException if error encountered during creationg of the file
```
Comment (albertshau) [inline twill-common/src/main/java/org/apache/twill/filesystem/Location.java]: typo: creation

File: twill-common/src/main/java/org/apache/twill/filesystem/Location.java [hunk 4]
```diff
@@ -53,11 +54,42 @@
...
    */
   boolean createNew() throws IOException;
 
   /**
+   * Atomically creates a new, empty file named by this abstract pathname if and only if a file with this name
+   * does not yet exist. The newly created file will have permission set based on the given permission settings.
+   *
+   * @param permission A permission string. It has to be either a three digits or a nine characters string.
+   *                   For the three digits string, it is similar to the UNIX permission numeric representation.
```
Comment (albertshau) [inline twill-common/src/main/java/org/apache/twill/filesystem/Location.java]: three digits -> three digit

File: twill-common/src/main/java/org/apache/twill/filesystem/Location.java [hunk 5]
```diff
@@ -53,11 +54,42 @@
...
+   *                   the third digit is the permission for all.
+   *                   For the nine characters string, it uses the format as specified by the
+   *                   {@link PosixFilePermissions#fromString(String)} method.
+   * @return {@code true} if the file is successfully create, {@code false} otherwise.
+   * @throws IOException if error encountered during creationg of the file
+   */
+  boolean createNew(String permission) throws IOException;
+
+  /**
+   * Returns the permissions of this {@link Location}. The permission string is a nine characters string as the format
```
Comment (albertshau) [inline twill-common/src/main/java/org/apache/twill/filesystem/Location.java]: nine characters -> nine character

File: twill-common/src/main/java/org/apache/twill/filesystem/Location.java [hunk 6]
```diff
@@ -73,7 +105,12 @@
    * Creates an {@link OutputStream} for this location with the given permission. The actual permission supported
    * depends on implementation.
    *
-   * @param permission A POSIX permission string.
+   * @param permission A permission string. It has to be either a three digits or a nine characters string.
+   *                   For the three digits string, it is similar to the UNIX permission numeric representation.
```
Comment (albertshau) [inline twill-common/src/main/java/org/apache/twill/filesystem/Location.java]: three digit, nine character

**Comment by albertshau** [comment]
only minor comments about javadocs, otherwise LGTM

File: twill-common/src/main/java/org/apache/twill/filesystem/LocalLocation.java (hunk 1 above)
Comment (chtyim) [inline twill-common/src/main/java/org/apache/twill/filesystem/LocalLocation.java]: fixed

File: twill-common/src/main/java/org/apache/twill/filesystem/LocalLocation.java (hunk 2 above)
Comment (chtyim) [inline twill-common/src/main/java/org/apache/twill/filesystem/LocalLocation.java]: fixed

File: twill-common/src/main/java/org/apache/twill/filesystem/Location.java (hunk 3 above)
Comment (chtyim) [inline twill-common/src/main/java/org/apache/twill/filesystem/Location.java]: fixed

File: twill-common/src/main/java/org/apache/twill/filesystem/Location.java (hunk 6 above)
Comment (chtyim) [inline twill-common/src/main/java/org/apache/twill/filesystem/Location.java]: fixed

File: twill-common/src/main/java/org/apache/twill/filesystem/Location.java (hunk 5 above)
Comment (chtyim) [inline twill-common/src/main/java/org/apache/twill/filesystem/Location.java]: fixed

File: twill-common/src/main/java/org/apache/twill/filesystem/Location.java (hunk 4 above)
Comment (chtyim) [inline twill-common/src/main/java/org/apache/twill/filesystem/Location.java]: fixed.

**Pull request closed without merging**